### PR TITLE
Fix TagHelper directive tests to fail when expected.

### PR DESCRIPTION
**Pull request by NTaylorMullen**
- Prior to this change the equals bits were falling back to the type comparison that `ChunkGenerator` exposes.

### CODE REVIEW

File: src/Microsoft.AspNetCore.Razor/Chunks/Generators/AddTagHelperChunkGenerator.cs [hunk 1]
```diff
@@ -35,5 +37,23 @@ public override void GenerateChunk(Span target, ChunkGeneratorContext context)
...
+            var other = obj as AddTagHelperChunkGenerator;
+            return base.Equals(obj) &&
+                string.Equals(_lookupText, other._lookupText, StringComparison.Ordinal);
+        }
+
+        /// <inheritdoc />
+        public override int GetHashCode()
+        {
+            var combiner = HashCodeCombiner.Start();
+            combiner.Add(base.GetHashCode());
```
Comment (ajaybhargavb) [inline src/Microsoft.AspNetCore.Razor/Chunks/Generators/AddTagHelperChunkGenerator.cs]: Why is this needed?

File: src/Microsoft.AspNetCore.Razor/Chunks/Generators/RemoveTagHelperChunkGenerator.cs [hunk 2]
```diff
@@ -35,5 +37,23 @@ public override void GenerateChunk(Span target, ChunkGeneratorContext context)
         {
             context.ChunkTreeBuilder.AddRemoveTagHelperChunk(_lookupText, target);
         }
+
+        /// <inheritdoc />
+        public override bool Equals(object obj)
+        {
+            var other = obj as RemoveTagHelperChunkGenerator;
+            return base.Equals(obj) &&
```
Comment (ajaybhargavb) [inline src/Microsoft.AspNetCore.Razor/Chunks/Generators/RemoveTagHelperChunkGenerator.cs]: Wouldn't checking `other != null` be sufficient?

File: src/Microsoft.AspNetCore.Razor/Chunks/Generators/AddTagHelperChunkGenerator.cs [hunk 3]
```diff
@@ -35,5 +37,23 @@ public override void GenerateChunk(Span target, ChunkGeneratorContext context)
         {
             context.ChunkTreeBuilder.AddAddTagHelperChunk(_lookupText, target);
         }
+
+        /// <inheritdoc />
+        public override bool Equals(object obj)
+        {
+            var other = obj as AddTagHelperChunkGenerator;
+            return base.Equals(obj) &&
+                string.Equals(_lookupText, other._lookupText, StringComparison.Ordinal);
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Razor/Chunks/Generators/AddTagHelperChunkGenerator.cs]: Couldn't `other` be null here?

Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Razor/Chunks/Generators/AddTagHelperChunkGenerator.cs]: the base.Equals ensures it's not `null`.

File: src/Microsoft.AspNetCore.Razor/Chunks/Generators/RemoveTagHelperChunkGenerator.cs (hunk 2 above)
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Razor/Chunks/Generators/RemoveTagHelperChunkGenerator.cs]: It would be functionally equivalent to what it is currently but by calling down to base we ensure that any sub-api equals changes are captured here.

File: src/Microsoft.AspNetCore.Razor/Chunks/Generators/AddTagHelperChunkGenerator.cs (hunk 1 above)
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Razor/Chunks/Generators/AddTagHelperChunkGenerator.cs]: https://github.com/aspnet/Razor/pull/753#discussion_r63648995

File: src/Microsoft.AspNetCore.Razor/Chunks/Generators/AddTagHelperChunkGenerator.cs (hunk 3 above)
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Razor/Chunks/Generators/AddTagHelperChunkGenerator.cs]: In the event `obj` is a `SpanChunkGenerator`  but not a `AddTagHelperChunkGenerator` this might be null. Change the first line to `base.Equals(other)` and you should be good.

Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Razor/Chunks/Generators/AddTagHelperChunkGenerator.cs]: ooo :+1: good catch

Comment (dougbu) [inline src/Microsoft.AspNetCore.Razor/Chunks/Generators/AddTagHelperChunkGenerator.cs]: This is fine due to the `GetType()` comparison in `base.Equals()`. But obviously it is harder to read than necessary.

**Comment by NTaylorMullen** [comment]
:up: :date:

**Comment by ajaybhargavb** [comment]
:shipit:

**Comment by pranavkm** [comment]
:shipit:
